politeiad/backendv2/tstorebe/tstore/tstore.go

```diff
@@ -0,0 +1,1288 @@
+
+		// Record index
+		m, err := merkleLeafHashForBlobEntry(*be)
+		if err != nil {
+			return nil, err
+		}
+		streamsIdx, ok := idx.Metadata[v.PluginID]
+		if !ok {
+			streamsIdx = make(map[uint32][]byte, len(metadata))
+		}
+		streamsIdx[v.StreamID] = m
+		idx.Metadata[v.PluginID] = streamsIdx
+
+		// Aggregate digest
+		digests[be.Digest] = struct{}{}
+	}
+
+	// Setup files
+	for _, v := range files {
+		// Blob entry
+		be, err := convertBlobEntryFromFile(v)
+		if err != nil {
+			return nil, err
+		}
+		beFiles[v.Name] = *be
+
+		// Record Index
+		m, err := merkleLeafHashForBlobEntry(*be)
+		if err != nil {
+			return nil, err
+		}
+		idx.Files[v.Name] = m
+
+		// Aggregate digest
+		digests[be.Digest] = struct{}{}
+	}
+
+	// Check if any of the content already exists. Different record
+	// versions that reference the same data is fine, but this data
+	// should not be saved to the store again. We can find duplicates
+	// by comparing the blob entry digest to the log leaf value. They
+	// will be the same if the record content is the same.
+	dups := make(map[string]struct{}, len(digests))
+	for _, v := range leavesAll {
+		d := hex.EncodeToString(v.LeafValue)
+		_, ok := digests[d]
+		if ok {
+			// A piece of the new record content already exsits in the
+			// tstore. Save the digest as a duplcate.
+			dups[d] = struct{}{}
+		}
+	}
+
+	// Prepare blobs for the kv store
+	var (
+		blobs  = make(map[string][]byte, len(digests))
+		leaves = make([]*trillian.LogLeaf, 0, len(blobs))
+
+		// dupBlobs contains the blob entries for record content that
+		// already exists. We may need these blob entries later on if
+		// the duplicate content is encrypted and it needs to be saved
+		// plain text.
+		dupBlobs = make(map[string]store.BlobEntry, len(digests))
+	)
+
+	// Only vetted data should be saved plain text
+	var encrypt bool
+	switch idx.State {
+	case backend.StateUnvetted:
+		encrypt = true
+	case backend.StateVetted:
+		// Save plain text
+		encrypt = false
+	default:
+		// Something is wrong
+		e := fmt.Sprintf("invalid record state %v %v", treeID, idx.State)
+		panic(e)
+	}
+
+	// Prepare record metadata blobs and leaves
+	_, ok := dups[beRecordMD.Digest]
+	if !ok {
+		// Not a duplicate. Prepare kv store blob.
+		b, err := store.Blobify(*beRecordMD)
+		if err != nil {
+			return nil, err
+		}
+		k := storeKeyNew(encrypt)
+		blobs[k] = b
+
+		// Prepare tlog leaf
+		extraData, err := extraDataEncode(k,
+			dataDescriptorRecordMetadata, idx.State)
+		if err != nil {
+			return nil, err
+		}
+		digest, err := hex.DecodeString(beRecordMD.Digest)
+		if err != nil {
+			return nil, err
+		}
+		leaves = append(leaves, newLogLeaf(digest, extraData))
+	} else {
+		// This is a duplicate. Stash is for now. We may need to save
+		// it as plain text later.
+		dupBlobs[beRecordMD.Digest] = *beRecordMD
+	}
+
+	// Prepare metadata stream blobs and leaves
+	for _, v := range beMetadata {
+		for _, be := range v {
+			_, ok := dups[be.Digest]
```

Comment:
Reverse logic for readability.

Reply:
Done 
https://github.com/lukebp/politeia/blob/tlog/politeiad/backendv2/tstorebe/tstore/record.go#L229